3.12.0b3 build issue: 'O_CLOEXEC' undeclared in sysmodule.c

I am building Python 3.12.0b3 (3.11.X builds fine) on CentOS 5 using glibc-2.5 and gcc-4.8.5 and I encounter this build error:

gcc-4.8 -std=gnu11 -pthread -c -fno-strict-overflow -Wsign-compare -DNDEBUG -g -O3 -Wall -std=c11 -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes -Werror=implicit-function-declaration -fvisibility=hidden  -I./Include/internal  -I. -I./Include   -fPIC -DPy_BUILD_CORE \
		-DABIFLAGS='""' \
		-DMULTIARCH=\"x86_64-linux-gnu\" \
		-o Python/sysmodule.o ./Python/sysmodule.c
./Python/sysmodule.c: In function 'PyUnstable_PerfMapState_Init':
./Python/sysmodule.c:2280:62: error: 'O_CLOEXEC' undeclared (first use in this function)
     int flags = O_WRONLY | O_CREAT | O_APPEND | O_NOFOLLOW | O_CLOEXEC;
                                                              ^
./Python/sysmodule.c:2280:62: note: each undeclared identifier is reported only once for each function it appears in
make: *** [Python/sysmodule.o] Error 1

Perhaps there is a missing #ifdef O_CLOEXEC in ./Python/sysmodule.c?

Thanks



<!-- gh-linked-prs -->
### Linked PRs
* gh-106120
* gh-106199
<!-- /gh-linked-prs -->

--- a/Misc/NEWS.d/next/Build/2023-06-26-21-56-29.gh-issue-106118.0cCfhl.rst
+++ b/Misc/NEWS.d/next/Build/2023-06-26-21-56-29.gh-issue-106118.0cCfhl.rst
@@ -1,0 +1,2 @@
+Fix compilation for platforms without :c:const:`!O_CLOEXEC`. The issue was
+introduced with Python 3.12b1 in :gh:`103295`. Patch by Erlend Aasland.

--- a/Misc/NEWS.d/next/Build/2023-06-26-21-56-29.gh-issue-106118.0cCfhl.rst
+++ b/Misc/NEWS.d/next/Build/2023-06-26-21-56-29.gh-issue-106118.0cCfhl.rst
@@ -1,2 +1,2 @@
-Fix compilation for platforms without :c:const:`!O_CLOEXEC`. The issue was
+Fix compilation for platforms without :data:`!O_CLOEXEC`. The issue was
 introduced with Python 3.12b1 in :gh:`103295`. Patch by Erlend Aasland.

--- a/Python/sysmodule.c
+++ b/Python/sysmodule.c
@@ -2277,7 +2277,10 @@ PyAPI_FUNC(int) PyUnstable_PerfMapState_Init(void) {
     char filename[100];
     pid_t pid = getpid();
     // Use nofollow flag to prevent symlink attacks.
-    int flags = O_WRONLY | O_CREAT | O_APPEND | O_NOFOLLOW | O_CLOEXEC;
+    int flags = O_WRONLY | O_CREAT | O_APPEND | O_NOFOLLOW;
+#ifdef O_CLOEXEC
+    flags |= O_CLOEXEC;
+#endif
     snprintf(filename, sizeof(filename) - 1, "/tmp/perf-%jd.map",
                 (intmax_t)pid);
     int fd = open(filename, flags, 0600);

